File: ANSv4/CatalogoLibraryVB/_devolucion/EnvioDevolucion.vb
Option Explicit On

Public Class EnvioDevolucion

    ' Define como se llama este modulo para el control de errores
    Private Const m_sMODULENAME_ As String = "clsEnvioDevolucion"



    ' Esta clase realiza una consulta al servidor web y obtiene
    ' el numero de ip del servidor privado.
    Private Cliente As DevolucionWS.Devolucion
    Private WebServiceInicializado As Boolean
    Private m_MacAddress As String
    Private m_ip As String

    Private m_NroDevolucion As String
    Private m_CodCliente As String
    Private m_Fecha As String
    Private m_Observaciones As String
    Private m_Detalle As String

    Private DatosObtenidos As Boolean

    Public Sub New(ByVal conexion As System.Data.OleDb.OleDbConnection, _
                   ByVal ipAddress As String, _
                   ByVal ipAddressIntranet As String, _
                   ByVal MacAddress As String, _
                   ByVal usaProxy As Boolean, _
                   ByVal proxyServerAddress As String)
        Inicializar(ipAddress, ipAddressIntranet, MacAddress,
                    usaProxy, proxyServerAddress)
        Conexion1 = conexion
    End Sub

    Public ReadOnly Property Inicializado() As Boolean
        Get
            Return WebServiceInicializado
        End Get
    End Property

    Private Conexion1 As OleDb.OleDbConnection

    Public Sub ObtenerDatos(ByVal NroDevolucion As String)

        DatosObtenidos = False

        Dim Enc As System.Data.OleDb.OleDbDataReader
        Dim Det As System.Data.OleDb.OleDbDataReader

        Enc = Funciones.adoModulo.adoComando(vg.Conexion, "EXECUTE v_Devolucion_Enc '" & NroDevolucion & "'")
        Det = Funciones.adoModulo.adoComando(vg.Conexion, "EXECUTE v_Devolucion_Det '" & NroDevolucion & "'")

        m_NroDevolucion = NroDevolucion
        m_CodCliente = Format(Trim(Enc("IDCliente")), "000000")
        m_Fecha = Enc("F_Devolucion")
        m_Observaciones = Replace(Enc("Observaciones"), ",", " ") & ""

        If Det.HasRows Then
            m_Detalle = ""
            Do While Det.Read
0:              m_Detalle = m_Detalle & Det("C_Producto") & ","                                '-00-
1:              m_Detalle = m_Detalle & Format(Trim(Det("Cantidad")), "00000000") & "00,"      '-01-
2:              m_Detalle = m_Detalle & "NO" & ","                                                '-02-
3:              m_Detalle = m_Detalle & "NO" & ","                                                '-03-
4:              m_Detalle = m_Detalle & "NO" & ","                                                '-04-
                '            'm_Detalle = m_Detalle & IIf(Len(Trim(rs_Det("miDeposito"))) = 0, "   ", rs_Det("miDeposito")) & ","                                '-05-
5:              m_Detalle = m_Detalle & Det("miDeposito") & ","
6:              m_Detalle = m_Detalle & Det("Factura") & ","                                   '-06-
7:              m_Detalle = m_Detalle & Format(Det("TipoDev"), "00") & ","                     '-07-
8:              m_Detalle = m_Detalle & Det("Vehiculo") & ","                                  '-08-
9:              m_Detalle = m_Detalle & Det("Modelo") & ","                                    '-09-
10:             m_Detalle = m_Detalle & Det("Motor") & ","                                     '-10-
11:             m_Detalle = m_Detalle & Det("Km") & ","                                        '-11-
12:             m_Detalle = m_Detalle & Det("Observaciones") & ";"                             '-12-
            Loop
        End If

        Enc = Nothing
        Det = Nothing

        DatosObtenidos = True

    End Sub

    Public Function EnviarDevolucion() As Long

        On Error GoTo ErrorHandler

        Dim Cancel As Boolean

        Cancel = False

        Dim resultado As Long

        If Not WebServiceInicializado Then
            Cancel = True
        End If

        If Not Cancel Then

            If Not (vg.TranActiva Is Nothing) Then
                vg.TranActiva.Rollback()
                vg.TranActiva = Nothing
            End If

            resultado = Cliente.EnviarDevolucion(m_MacAddress, m_NroDevolucion, m_CodCliente, m_Fecha, m_Observaciones, m_Detalle)

            If resultado = 0 Then
                '                adoModulo.adoComandoIU(vg.Conexion, "EXEC usp_Devolucion_Transmicion_Upd '" & m_NroDevolucion & "'")
                If Not (vg.TranActiva Is Nothing) Then
                    vg.TranActiva.Commit()
                    vg.TranActiva = Nothing
                End If
            Else
                If Not (vg.TranActiva Is Nothing) Then
                    vg.TranActiva.Rollback()
                    vg.TranActiva = Nothing
                End If
            End If

            EnviarDevolucion = resultado
        Else
            EnviarDevolucion = -1
        End If

        Exit Function

ErrorHandler:
        Err.Raise(Err.Number, Err.Source, Err.Description)

    End Function

    Public Sub Inicializar(ByVal ipAddress As String, _
                           ByVal ipAddressIntranet As String, _
                           ByVal MacAddress As String, _
                           ByVal usaProxy As Boolean, _
                           ByVal proxyServerAddress As String)

        Dim Conectado As Boolean
        Conectado = My.Computer.Network.Ping(ipAddress, 5000)
        Conectado = True
        If Not Conectado Then
            Conectado = My.Computer.Network.Ping(ipAddressIntranet, 5000)
        End If

        On Error GoTo ErrorHandler

        If Conectado Then
            If Not WebServiceInicializado Then
                Cliente = New DevolucionWS.Devolucion
                Cliente.Url = "http://" & ipAddress & "/wsCatalogo3/Devolucion.asmx?wsdl"
                If usaProxy Then
                    Cliente.Proxy = New System.Net.WebProxy(proxyServerAddress)
                End If

                m_MacAddress = MacAddress
                m_ip = ipAddress
                WebServiceInicializado = True
            End If
        Else
            WebServiceInicializado = False
        End If

        Exit Sub

ErrorHandler:

        If Err.Number = -2147024809 Then
            ' Intento con el ip interno
            Cliente = New DevolucionWS.Devolucion
            Cliente.Url = "http://" & ipAddressIntranet & "/wsCatalogo3/Devolucion.asmx?wsdl"
            If usaProxy Then
                Cliente.Proxy = New System.Net.WebProxy(proxyServerAddress)
            End If

            m_MacAddress = MacAddress
            m_ip = ipAddressIntranet
            WebServiceInicializado = True
            Err.Clear()
        Else
            Err.Raise(Err.Number, Err.Source, Err.Description)
        End If

    End Sub

End Class

File: ANSv4/Catalogo4/_devolucion/EnvioDevolucion.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;
using System.Data;
using System.Diagnostics;
//using Microsoft.VisualBasic;

namespace Catalogo._devolucion
{
    public class EnvioDevolucion
    {

        // Esta clase realiza una consulta al servidor web y obtiene
        // el numero de ip del servidor privado.
        private DevolucionWS.Devolucion Cliente;
        private bool WebServiceInicializado;
        private string m_MacAddress;

        private string m_ip;
        private string m_NroDevolucion;
        private string m_CodCliente;
        private string m_Fecha;
        private string m_Observaciones;

        private string m_Detalle;

        private bool DatosObtenidos;
        public EnvioDevolucion(System.Data.OleDb.OleDbConnection conexion, string ipAddress, string ipAddressIntranet, string MacAddress, bool usaProxy, string proxyServerAddress)
        {
            Inicializar(ipAddress, ipAddressIntranet, MacAddress, usaProxy, proxyServerAddress);
            Conexion1 = conexion;
        }

        public bool Inicializado
        {
            get { return WebServiceInicializado; }
        }


        private System.Data.OleDb.OleDbConnection Conexion1;

        public void ObtenerDatos(string NroDevolucion)
	{
		DatosObtenidos = false;

		System.Data.OleDb.OleDbDataReader Enc = null;
		System.Data.OleDb.OleDbDataReader Det = null;

		Enc = Funciones.oleDbFunciones.Comando(Conexion1, "EXECUTE v_Devolucion_Enc '" + NroDevolucion + "'");
		Det = Funciones.oleDbFunciones.Comando(Conexion1, "EXECUTE v_Devolucion_Det '" + NroDevolucion + "'");

		m_NroDevolucion = NroDevolucion;
		m_CodCliente = String.Format("000000", Enc["IDCliente"].ToString().Trim());
		m_Fecha = Enc["F_Devolucion"].ToString();
		m_Observaciones = Enc["Observaciones"].ToString().Replace(",", " ");

		if (Det.HasRows) 
        {
			m_Detalle = "";
			while (Det.Read()) 
            {
				m_Detalle += Det["C_Producto"].ToString() + ",";
				m_Detalle += String.Format("00000000", Det["Cantidad"].ToString().Trim()) + "00,";
				m_Detalle += "NO" + ",";
				m_Detalle += "NO" + ",";
				m_Detalle += "NO" + ",";
				m_Detalle += Det["miDeposito"].ToString() + ",";
				m_Detalle += Det["Factura"].ToString() + ",";
				m_Detalle += String.Format("00", Det["TipoDev"]) + ",";
				m_Detalle += Det["Vehiculo"].ToString() + ",";
				m_Detalle += Det["Modelo"].ToString() + ",";
				m_Detalle += Det["Motor"].ToString() + ",";
				m_Detalle += Det["Km"].ToString() + ",";
				m_Detalle += Det["Observaciones"].ToString() + ";";
			}
		}

		DatosObtenidos = true;
	}

        public long EnviarDevolucion()
        {
            long functionReturnValue = 0;

            bool Cancel = false;

            long resultado = 0;

            if (!WebServiceInicializado)
            {
                Cancel = true;
            }


            if (!Cancel)
            {
            //    if ((vg.TranActiva != null))
              //  {
                //    vg.TranActiva.Rollback();
              //      vg.TranActiva = null;
                //}

                resultado = Cliente.EnviarDevolucion(m_MacAddress, m_NroDevolucion, m_CodCliente, m_Fecha, m_Observaciones, m_Detalle);

                if (resultado == 0)
                {
                    //                adoModulo.adoComandoIU(vg.Conexion, "EXEC usp_Devolucion_Transmicion_Upd '" & m_NroDevolucion & "'")
//                    if ((vg.TranActiva != null))
//                    {
//                        vg.TranActiva.Commit();
//                        vg.TranActiva = null;
//                    }
                }
                else
                {
//                    if ((vg.TranActiva != null))
//                    {
//                        vg.TranActiva.Rollback();
//                        vg.TranActiva = null;
//                   }
                }

                return resultado;
            }
            else
            {
                return -1;
            }
        }


        public void Inicializar(string ipAddress, string ipAddressIntranet, string MacAddress, bool usaProxy, string proxyServerAddress)
        {
            bool Conectado = util.SimplePing.ping(ipAddress, 5000);
            if (!Conectado)
            {
                Conectado = util.SimplePing.ping(ipAddressIntranet, 5000);
            }
            try
            {
                if (Conectado)
                {
                    if (!WebServiceInicializado)
                    {
                        Cliente = new DevolucionWS.Devolucion();
                        Cliente.Url = "http://" + ipAddress + "/wsCatalogo4/Devolucion.asmx?wsdl";
                        if (usaProxy)
                        {
                            Cliente.Proxy = new System.Net.WebProxy(proxyServerAddress);
                        }

                        m_MacAddress = MacAddress;
                        m_ip = ipAddress;
                        WebServiceInicializado = true;
                    }
                }
                else
                {
                    WebServiceInicializado = false;
                }
            }
            catch
            {
                //            if (Err().Number == -2147024809)
                {
                    // Intento con el ip interno
                    Cliente = new DevolucionWS.Devolucion();
                    Cliente.Url = "http://" + ipAddressIntranet + "/wsCatalogo4/Devolucion.asmx?wsdl";
                    if (usaProxy)
                    {
                        Cliente.Proxy = new System.Net.WebProxy(proxyServerAddress);
                    }

                    m_MacAddress = MacAddress;
                    m_ip = ipAddressIntranet;
                    WebServiceInicializado = true;
                }
                //          else
                //          {
                //              Err().Raise(Err().Number, Err().Source, Err().Description);
                //          }
            }
        }

    }
}